File: medistock-backend/src/main/java/com/medistock/service/MedicineService.java
package com.medistock.service;

import com.medistock.entity.Medicine;
import com.medistock.repository.MedicineRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
public class MedicineService {

    @Autowired
    private MedicineRepository medicineRepository;

    @Autowired
    private InventoryHistoryService historyService;

    // ===========================
    // Add Medicine
    // ===========================
    public Medicine addMedicine(Medicine medicine) {

        Medicine savedMedicine = medicineRepository.save(medicine);

        historyService.saveHistory(
                savedMedicine.getMedicineName(),
                "Added",
                savedMedicine.getQuantity()
        );

        return savedMedicine;
    }

    // ===========================
    // Get All Medicines
    // ===========================
    public List<Medicine> getAllMedicines() {
        return medicineRepository.findAll();
    }

    // ===========================
    // Get Medicine By Id
    // ===========================
    public Medicine getMedicine(Long id) {
        return medicineRepository.findById(id).orElse(null);
    }

    // ===========================
    // Update Medicine
    // ===========================
    public Medicine updateMedicine(Long id, Medicine medicine) {

        Medicine existing = medicineRepository.findById(id).orElse(null);

        if (existing == null) {
            return null;
        }

        existing.setMedicineName(medicine.getMedicineName());
        existing.setCategory(medicine.getCategory());
        existing.setBatchNumber(medicine.getBatchNumber());
        existing.setQuantity(medicine.getQuantity());
        existing.setPrice(medicine.getPrice());
        existing.setManufacturingDate(medicine.getManufacturingDate());
        existing.setExpiryDate(medicine.getExpiryDate());

        Medicine updatedMedicine = medicineRepository.save(existing);

        historyService.saveHistory(
                updatedMedicine.getMedicineName(),
                "Updated",
                updatedMedicine.getQuantity()
        );

        return updatedMedicine;
    }

    // ===========================
    // Delete Medicine
    // ===========================
    public void deleteMedicine(Long id) {

        Medicine medicine = medicineRepository.findById(id).orElse(null);

        if (medicine != null) {

            historyService.saveHistory(
                    medicine.getMedicineName(),
                    "Deleted",
                    medicine.getQuantity()
            );

            medicineRepository.deleteById(id);
        }
    }

    // ===========================
    // Search By Name
    // ===========================
    public List<Medicine> searchMedicine(String name) {
        return medicineRepository.findByMedicineNameContainingIgnoreCase(name);
    }

    // ===========================
    // Low Stock
    // ===========================
    public List<Medicine> getLowStockMedicines() {
        return medicineRepository.findByQuantityLessThan(10);
    }

    // ===========================
    // Out Of Stock
    // ===========================
    public List<Medicine> getOutOfStockMedicines() {
        return medicineRepository.findByQuantity(0);
    }

    // ===========================
    // Near Expiry
    // ===========================
    public List<Medicine> getNearExpiryMedicines() {

        LocalDate today = LocalDate.now();
        LocalDate next30Days = today.plusDays(30);

        return medicineRepository.findByExpiryDateBetween(today, next30Days);
    }

    // ===========================
    // Expired Medicines
    // ===========================
    public List<Medicine> getExpiredMedicines() {
        return medicineRepository.findByExpiryDateBefore(LocalDate.now());
    }

    // ===========================
    // Search By Category
    // ===========================
    public List<Medicine> searchByCategory(String category) {
        return medicineRepository.findByCategoryContainingIgnoreCase(category);
    }

    // ===========================
    // Search By Batch
    // ===========================
    public List<Medicine> searchByBatch(String batch) {
        return medicineRepository.findByBatchNumberContainingIgnoreCase(batch);
    }

    // ===========================
    // Search By Expiry Date
    // ===========================
    public List<Medicine> searchByExpiry(LocalDate expiryDate) {
        return medicineRepository.findByExpiryDate(expiryDate);
    }

}

File: medistock-frontend/src/services/medicineService.js
import axios from "axios";

const API = "http://localhost:8081/api/medicines";

const authConfig = () => {

    const token = localStorage.getItem("token");

    return {
        headers: {
            Authorization: `Bearer ${token}`
        }
    };

};

export const getMedicines = () =>
    axios.get(API, authConfig());

export const addMedicine = (medicine) =>
    axios.post(API, medicine, authConfig());

export const updateMedicine = (id, medicine) =>
    axios.put(
        `${API}/${id}`,
        medicine,
        authConfig()
    );

export const deleteMedicine = (id) =>
    axios.delete(
        `${API}/${id}`,
        authConfig()
    );

export const searchMedicine = (name) =>
    axios.get(
        `${API}/search?name=${encodeURIComponent(name)}`,
        authConfig()
    );

export const getLowStockMedicines = () =>
    axios.get(
        `${API}/low-stock`,
        authConfig()
    );

export const getOutOfStockMedicines = () =>
    axios.get(
        `${API}/out-of-stock`,
        authConfig()
    );

export const getNearExpiryMedicines = () =>
    axios.get(
        `${API}/near-expiry`,
        authConfig()
    );

export const getExpiredMedicines = () =>
    axios.get(
        `${API}/expired`,
        authConfig()
    );

export const searchByCategory = (category) =>
    axios.get(
        `${API}/search/category?category=${encodeURIComponent(category)}`,
        authConfig()
    );

export const searchByBatch = (batch) =>
    axios.get(
        `${API}/search/batch?batch=${encodeURIComponent(batch)}`,
        authConfig()
    );

export const getMedicineCount = () =>
    axios.get(
        `${API}/count`,
        authConfig()
    );

export const downloadExcel = () => {

    const token = localStorage.getItem("token");

    window.open(
        `${API}/excel?token=${token}`
    );

};

export const downloadPdf = () => {

    const token = localStorage.getItem("token");

    window.open(
        `${API}/pdf?token=${token}`
    );

};